### PR TITLE
Move `addon-info` into addons panel

### PR DESCRIPTION
**Pull request by usulpro**
Issue: #1147

## What I did

image

`withInfo()` puts info into Info panel

`decoratorInfo()` keeps previous behavior

## How to test

run `cra-kitchen-sink`

## Todo

- [ ] check out https://github.com/storybooks/storybook/labels/addon%3A%20info

- [x] find more complex examples to test Info Addon with full load

- [x] implement #811 for better stories organization

- [ ] add tests

- [x] options clean-up

- [ ] add full documentation

- [ ] move some UI into `lib/components`

- [x] restore markdown support

### CODE REVIEW

**Comment by shilman** [comment]
Awesome!! So excited to see the progress on this and I think it will be a huge benefit for Storybook! 🎉 

A few thoughts:

1. What does #811 have to do with this? I think it's unrelated and better to keep it in a separate PR, but maybe I don't understand something.
2. Can we please remove (or at least deprecate) the inline support? I think it's an anti-pattern for Storybook because (1) it messes up testing, and (2) it makes the addon react-specific (unnecessarily), (3) it complicates the code to support two modes

**Comment by ls-guillaume-lambert** [comment]
Quick comment on the styling of the Info panel: I wouldn't make it so "boxy", and keep the padding to the minimum to take as much screen size as possible. Is this planned for an upcoming release?

**Comment by alterx** [comment]
+1 to deprecate (or, better yet, remove) inline support. It's one of the things that makes this (and potentially other) add-on difficult to use with, Angular, for example.

**Comment by ndelangen** [comment]
I think this is an opportunity to move some UI into `lib/components` !

**Comment by UsulPro** [comment]
- [x] remove decorator


- [x] update styles to fit small screens


- [x] rename setDefaults to setInfoOptions 

<details>
  <summary>
 Reasons:
  </summary>

1. Setting options for an addon is a good idea and we might want to include it to `withX` standardization so I want that name to be relevant to addon's name.

2. This function could be used not only to initiate setting so I remove 'default' from name

3. The use cases are very similar to `setOptions` 

</details>
.

- [x] `setInfoOptions`, `defaultOptions` and `withInfo` should support the same set of options. `setInfoOptions` and `withInfo` has same interface related to `textOrOptions`

- [x] rename `options.text` to `options.info`

<details>
  <summary>
 Reasons:
  </summary>

1. It supports not only plain text but markdown as well so `text` could confuse users.

2. It's a "leading" prop and could be passed directly without `{options}` so name `info` could highlight it as a main.

3. Same as an inner prop of `Story`

counter-argument: since we publish v3.2 it will be a "breaking" change. So I don't mind to revert it to the previous name.

</details>
.

### How to test

Check different options of "Addon info" from `cra-kitchen-sink`

**Comment by UsulPro** [comment]
Known issues at this moment:

- [ ] Shows "No info provided..." after switching to another addon panel

- [ ] Poor markdown support

**Comment by shilman** [comment]
@UsulPro how about instead of `options.text` => `options.markdown`?

**Comment by UsulPro** [comment]
@shilman This name clearly reflects the type of data. But on the other hand, all other props are named in accordance with their purpose. e.g.: `header`, `source` so I'm not sure that it would fit the overall style. I guess it could be something like `description` from this point of view.
But of course, let's choose what is most convenient for users

And why you don't like `info` name? We could do the same for `addon-notes`: set `notes` as the "main" prop and use it in a "short" version.

**Comment by shilman** [comment]
@UsulPro i have a proposal for an updated addons API and I think it becomes confusing if we have the main property have the same name as the addon:

https://gist.github.com/shilman/792dc25550daa9c2bf37238f4ef7a398#options-types

We haven't discussed this proposal, and there might be other better proposals out there. But if the proposal goes through then it would be better to call it anything other than `info` IMHO (`markdown`, `summary`, etc.) because `info.summary` makes sense but `info.info` is confusing. LMK what you think.

**Comment by UsulPro** [comment]
@shilman Let's take `summary`, it looks nice to me!

**Comment by UsulPro** [comment]
Previous versions of Addon Info supported JSX syntax for the main argument. At this moment "summary" supports JSX only when it passed via `{options}` e.g:

```js
.add(
    'info with JSX description',
    withInfo({
      summary: (<div>custom JSX structure</div>),
      header: false,
      source: false,
    })(<button> Press Me </button>)
)

``` 

but it doesn't work if `summary` is an only argument.

@shilman I'm not sure do we still need to support this option in a single argument form? Maybe string will be enough and {options} for all other more complicated cases?

**Comment by UsulPro** [comment]
let's support it for a while

**Comment by ls-guillaume-lambert** [comment]
Question: is this going to be optional? I like having the info inside the story and not hidden in a panel 😉

**Comment by UsulPro** [comment]
@ls-guillaume-lambert I guess for the first time we left both options
But we aren't going to support decorator form and will remove it completely sooner or later
you can find the discussion here: #1147 

BTW: In Storybook v3.2.4 was introduced the new option to select a specific addon panel via `setOptions`: `selectedAddonPanel`. You can use it per story to select appropriate panel and keep your info always opened
https://github.com/storybooks/storybook/tree/master/addons/options#getting-started

**Comment by UsulPro** [comment]
I added Storybook to `addon-info` with more examples and own components

```
cd addons/info
yarn dev
```

Hope it'll help with testing and development in the future

**Comment by Hypnosphi** [comment]
> do not do `date('Label', new Date())` or `date('Label')

Can you please move it to `cra-kitchen-sink` (as @ndelangen did with stories for `lib/components`?

**Comment by Hypnosphi** [comment]
Oops, wrong copypaste =D

I meant this:
> I added Storybook to addon-info with more examples and own components

**Comment by UsulPro** [comment]
@Hypnosphi I need to clarify:
I added examples of `addon-info` to `cra-kitchen-sink` which represented the most common use cases
**and** added own storybook with internal components to use in the development mode

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 1]
```diff
@@ -108,7 +108,8 @@ storiesOf('Button', module)
     const nice = boolean('Nice', true);
 
     // NOTE: put this last because it currently breaks everything after it :D
```
Comment (ndelangen) [inline examples/cra-kitchen-sink/src/stories/index.js]: I think this comment can be removed right?

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: addons/info/src/components/markdown/index.js [hunk 2]
```diff
@@ -1,3 +1,6 @@
-export { H1, H2, H3, H4, H5, H6 } from './htags';
-export { Code, Pre } from './code';
-export { P, Small, A, LI } from './text';
+import { getOptions } from './compiler';
```
Comment (ndelangen) [inline addons/info/src/components/markdown/index.js]: Same file is imported twice

File: addons/info/src/components/stories/index.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
+/* eslint-disable react/prop-types */
+
+import React from 'react';
```
Comment (ndelangen) [inline addons/info/src/components/stories/index.js]: Move examples to the cra-kitchen-sink

File: addons/info/src/components/stories/mock-data/summary.md.js [hunk 4]
```diff
@@ -0,0 +1,300 @@
+export default `
+
```
Comment (ndelangen) [inline addons/info/src/components/stories/mock-data/summary.md.js]: That is some regression test, I like it!

File: addons/info/src/index.js [hunk 5]
```diff
@@ -1,74 +1,116 @@
...
-  return (
+
+  const infoContent = (
     <Story {...props}>
-      {storyFn(context)}
+      {story}
     </Story>
   );
+  if (options.sendToPanel) {
+    const infoString = ReactDOMServer.renderToString(infoContent);
```
Comment (ndelangen) [inline addons/info/src/index.js]: This renerToString nessesary here? Why? 

`ReactDOMServer` is for server usage right? Won't it blow the manager.bundle size out of the park?

File: addons/info/src/components/stories/mock-data/widget.jsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
```
Comment (ndelangen) [inline addons/info/src/components/stories/mock-data/widget.jsx]: A component within mock-data? Sounds like this folder should contain only data ( js|json )
?

File: addons/info/src/panel/empty_panel.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+
+const styles = {
+  main: {
+    fontFamily:
+      '-apple-system, ".SFNSText-Regular", "San Francisco", Roboto, "Segoe UI", "Helvetica Neue", "Lucida Grande", sans-serif',
```
Comment (ndelangen) [inline addons/info/src/panel/empty_panel.js]: Use theme from lib/components, or better yet, don't try to change to another font.

Preferably do not override the font at all.

File: addons/info/src/panel/empty_panel.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  return (
+    <div style={styles.main}>
+      INFO PANEL
+      <p>No info provided for this story</p>
+      <p>
+        Use <code style={styles.code}>withInfo()</code> to add info addon
+      </p>
+      <p>
+        See{' '}
+        <a href="https://github.com/storybooks/storybook/tree/master/addons/info">
```
Comment (ndelangen) [inline addons/info/src/panel/empty_panel.js]: Nice!

Comment (ndelangen) [inline addons/info/src/panel/empty_panel.js]: Although I eventually would only want to show panels that actually have meaning for a story. (far future stuff)

File: addons/info/src/panel/index.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import EmptyPanel from './empty_panel';
+
+/* select one from src/components/markdown/hljs-styles/
```
Comment (ndelangen) [inline addons/info/src/panel/index.js]: We should either only include one, or allow a consumer to inject their own.

Having a whole bunch of themes in our codebase no-one can use unless they change OUR code, makes little sense, or am I reading this wrong?

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: addons/info/src/components/Node.js [hunk 10]
```diff
@@ -5,7 +5,7 @@ import Props from './Props';
 const stylesheet = {
   containerStyle: {},
   tagStyle: {
-    color: '#777',
+    color: 'hsl(0, 0%, 10%)',
```
Comment (Hypnosphi) [inline addons/info/src/components/Node.js]: Let's use `hsl` for grayscale colors in #1360 as well

**Comment by UsulPro** [comment]
## Glamorous

Some review comments are related to moving components into `lib/components` and using glamorous. Yes, I kept it in mind while doing it and I think some components of this package are good candidates to be moved. I didn't move them yet only because this PR is already pretty big, and I'd like to keep it located only within this package, to make review easier. So after we finish with this PR, I'd like to create some small ones to complete all remaining tasks, if you don't mind.

I think one of the possible components which we'd like to have in `lib/components` package could be one to show code source with syntax highlighting. I have made some effort to make it ~great~ work again on the panel side. Now it displays fenced code sections in markdown. So what do you think about to reuse it to show the story source code in the "Story Source" section? We need to implement this #1602 before to make it possible.

This component uses (~highlight.js~ -> [lowlight](https://github.com/wooorm/lowlight)) wich have [82 own css styles](https://github.com/isagalaev/highlight.js/tree/master/src/styles). At this moment I just import one globally, but as @ndelangen noted, this is not a good solution. I think we need to decide do we want to allow for users to choose own styles or not. If not, I'll just convert the default one (`github.css`) to glamorous. Otherwise, I'd like you to share ideas how to better deal with them?

**Comment by UsulPro** [comment]
## Performance

Other notes are related to performance issues. Yes, there are some, you can find them while switching between stories: if each is used with addon-info you can see the short "empty panel" blinked before new info will appear. Though it doesn't look critical, we can make it better and get abilities for the future addon improvements.

#### Few words of how it works now:
Before this PR we just returned a new React Component which gets our story as a prop. We can't pass React Component through the channel directly, so I render it to the string via `ReactDOMServer.renderToString`, send to the channel and then insert it via `dangerouslySetInnerHTML`. Obviously, it's not the optimal way to do. We can divide it into chunks by types and process on the panel side e.g. we can send markdown as a string instead of compiling it into React stringified code before sending. So I guess it'll change the logic a bit and some issues will be solved.  But I think some parts related to the story itself we'll still need to render on the preview side. So do you have any suggestions how to avoid `ReactDOMServer.renderToString` in this case?

**Comment by Hypnosphi** [comment]
I'm strongly opposed to having lots of third-party css as part of our codebase

**Comment by UsulPro** [comment]
## Storybook

I see you all ask me to move Storybook into cra-kitchen-sink. I still not sure it's a good idea, but let me please explain why I think so.

First of all, in this PR I already add a lot of `addon-info` examples to cra-kitchen-sink. Actually, I created a separate `storyKind` for addon-info and if you run it you'll find out a bunch of stories representing different use cases. So I more wonder why nobody asked me to reduce it? (Maybe we'll do it after settling down with addon-refactoring?) It's because `addon-info` has a lot of features some of them aren't documented and I just dug it out while reviewing the code. In order to explore these features and own addon-info components, I ~created~ recreated the Storybook. Actually, it already had the `example` folder with a storybook, I removed it and some unused code as a clean-up. Then I found this storybook with internal components very useful for development small parts of code in isolation. (note that I use here addons `info` and `knobs` as a good tool to work with these components, but **not** as addon examples). So it's the very different Storybook containing internal things. Since we have a lot of issues/PRs related to different parts of addon-info I'd suggest to use it as a dev tool and ask contributors to reproduce bug/features with it.

Perhaps along with the unit tests, it might be a useful thing for keeping code in good condition? While cra-kitchen-sink is more like a showcase representing the finished products. Nothing prevents to run both at the same time.

Do you still want to union these storybooks or just to remove the internal one? I'm not religious about it of course and open for further discussion :smile:

File: addons/info/.storybook/addons.js [hunk 11]
```diff
@@ -1 +1,5 @@
+import '@storybook/addon-info/register';
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/.storybook/addons.js]: Please, see my [comment](https://github.com/storybooks/storybook/pull/1501#issuecomment-325218933) below

File: addons/info/package.json [hunk 12]
```diff
@@ -10,26 +10,38 @@
   },
   "scripts": {
     "prepublish": "node ../../scripts/prepublish.js",
-    "publish-storybook": "bash .scripts/publish_storybook.sh",
-    "storybook": "start-storybook -p 9010"
+    "dev": "start-storybook -p 9009",
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/package.json]: https://github.com/storybooks/storybook/pull/1501#issuecomment-325218933

File: addons/info/src/components/PropTable.js [hunk 13]
```diff
@@ -15,9 +15,14 @@ Object.keys(PropTypes).forEach(typeName => {
 
 const stylesheet = {
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/PropTable.js]: Yes, it's in my plans. There some questions I'd like to discuss before, see in the https://github.com/storybooks/storybook/pull/1501#issuecomment-325218127

File: addons/info/src/components/Story.js [hunk 14]
```diff
@@ -5,11 +5,11 @@ import PropTypes from 'prop-types';
 import global from 'global';
 import { baseFonts } from '@storybook/components';
 
-import marksy from 'marksy';
+import { compile, setOptions, pre as Pre } from './markdown';
+import './markdown/hljs-styles/androidstudio.css';
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/Story.js]: Absolutely agree! Maybe it's more related to components itself. Here it's just for development. I'd like to do it via glamorous, let's decide what API we wont to have?

File: addons/info/src/components/Story.js [hunk 15]
```diff
@@ -240,9 +257,14 @@ export default class Story extends React.Component {
...
     const source = lines.map(s => s.slice(padding)).join('\n');
+
+    setOptions({
+      components: this.props.components,
+    });
+
     return (
       <div style={this.state.stylesheet.infoContent}>
-        {this.marksy(source).tree}
+        {compile(source)}
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/Story.js]: Ok, I see! I can change it now, but maybe it worth to finish after solving other performance issues explained here: https://github.com/storybooks/storybook/pull/1501#issuecomment-325218401

File: addons/info/src/components/markdown/compiler.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    })
+    .use(remarkEmoji)
+    .use(remarkHighlight)
+    .use(remarkReact, { remarkReactComponents: options.components, sanitize: false });
+}
+
+let processor = setupProcessor();
+
+export function setOptions(newOptions) {
+  options = {
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/markdown/compiler.js]: Ok I will look for a better solution

File: addons/info/src/components/markdown/defaults.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    backgroundColor: 'hsl(0, 0%, 100%)',
+  },
+  code: {
+    textColor: 'hsl(218, 54%, 20%)',
+    backgroundColor: 'hsl(0, 0%, 96%)',
+    borderColor: 'hsl(0, 0%, 80%)',
+    borderRadius: 3,
+  },
+  baseFonts:
+    '-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen,Ubuntu,"Fira Sans","Droid Sans","Helvetica Neue",sans-serif',
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/markdown/defaults.js]: Ok

File: addons/info/src/components/markdown/hljs-styles/androidstudio.css [hunk 18]
```diff
@@ -0,0 +1,66 @@
+/*
```
[the remark this answers is not in the document]
Comment (UsulPro) [reply addons/info/src/components/markdown/hljs-styles/androidstudio.css]: I've take these style from https://github.com/isagalaev/highlight.js/tree/master/src/styles. A few commits ago it was much more, I just left some here for dev/review purposes. I guess `github.css` could fit for us as a default style, but might we wont to choose another one? After that I'll remove unused or we can think about API to allow to change them? See explanation in this https://github.com/storybooks/storybook/pull/1501#issuecomment-325218127

File: addons/info/src/components/markdown/index.js (hunk 2 above)
Comment (UsulPro) [reply addons/info/src/components/markdown/index.js]: will fix it

File: addons/info/src/components/stories/index.js (hunk 3 above)
Comment (UsulPro) [reply addons/info/src/components/stories/index.js]: I don't have examples in this storybook. It's more to have internal components in isolation and to test them with mock data. Do we need all this stuff in cra-kitchen-sink?
https://github.com/storybooks/storybook/pull/1501#issuecomment-325218933

File: addons/info/src/components/stories/mock-data/summary.md.js (hunk 4 above)
Comment (UsulPro) [reply addons/info/src/components/stories/mock-data/summary.md.js]: Thanks)

File: addons/info/src/index.js (hunk 5 above)
Comment (UsulPro) [reply addons/info/src/index.js]: It's a good question! I'm not sure how can we get this without `ReactDOMServer`, but maybe you have ideas? More details why it is: https://github.com/storybooks/storybook/pull/1501#issuecomment-325218401

File: addons/info/src/components/stories/mock-data/widget.jsx (hunk 6 above)
Comment (UsulPro) [reply addons/info/src/components/stories/mock-data/widget.jsx]: mmm... why a component can't be a mock-data? `addon-info` can take a React Component as an argument to show in the description section (now it's "summary"). I just need something to pass as an example to see how it works.

Comment (UsulPro) [reply addons/info/src/components/stories/mock-data/widget.jsx]: But possible I already don't use it, I need to check actually

File: addons/info/src/panel/empty_panel.js (hunk 7 above)
Comment (UsulPro) [reply addons/info/src/panel/empty_panel.js]: Ok!

File: addons/info/src/panel/index.js (hunk 9 above)
Comment (UsulPro) [reply addons/info/src/panel/index.js]: Is there any easy way to do it?
Actually, I should remove it!

File: addons/info/src/components/stories/index.js (hunk 3 above)
Comment (ndelangen) [reply addons/info/src/components/stories/index.js]: Yes i think it's better to put the edge cases of a addon in a story inside the kitchen sink:
```js
storiesOf('addons/info',module).add('usecase of info', () => {});
```

**Comment by danielduan** [comment]
Closing this because there hasn't been any activity here for a while. Feel free to reopen when this is ready for discussion again.

**Comment by ndelangen** [comment]
😢

**Comment by danielduan** [comment]
@Hypnosphi I agree that this should be worked on but I closed this because there is a significant amount of work left and there hasn't been any progress here since almost 3 months ago.

If you'd like to take over the PR, please do. I think you can branch off of this and create a new PR.

**Comment by Hypnosphi** [comment]
I think it’s a thing that really needs to be done, but I’m not sure I’m able to do it properly. What should I do in that case?

**Comment by danielduan** [comment]
I don't think there is a right person for this but @UsulPro definitely has more experience working in info than anyone else. I think the difficult part is coming to a consensus as to what the future info panel will look like and how it will function, much of which has been explored here.

I think what's here is a very good start. The rest of the work looks like it's fixing merge conflicts and integrating the changes with ongoing development since the last commit here. Migrating the Storybook over shouldn't be too difficult. We can then discuss if the changes here are appropriate.

**Comment by igor-dv** [comment]
Does anyone want to finish it for the v4 release?

**Comment by igor-dv** [comment]
I think we don't need to hold this one open. Let's revisit the implementation/rationale in the future.

**Pull request closed without merging**